File: python-backend/chart_spec.py
"""ChartSpec JSON-Schema dump.

The TS side mirrors the ChartSpec union in ``src/shared/chartSpec.ts``.
A drift check (``src/test/chartSpecSchema.test.ts``) spawns this module
as ``python -m chart_spec`` and diffs the output against a checked-in
snapshot, so the two languages stay in lockstep without a code
generator.

Each ``oneOf`` branch covers one ``kind``:

* ``line``    — x/y line plot, one or more series.
* ``heatmap`` — 2D matrix indexed by xs and ys.
* ``bar``     — categorical bars, one or more series.
* ``scalar``  — small label/value/unit table.
"""

from __future__ import annotations

import json
import sys


def dump_schema() -> dict:
    """Return the Draft 2020-12 JSON Schema describing the ChartSpec union."""
    line = {
        "type": "object",
        "required": ["kind", "title", "xLabel", "yLabel", "xs", "series"],
        "properties": {
            "kind": {"const": "line"},
            "title": {"type": "string"},
            "xLabel": {"type": "string"},
            "yLabel": {"type": "string"},
            "xs": {"type": "array", "items": {"type": "number"}},
            "series": {
                "type": "array",
                "items": {
                    "type": "object",
                    "required": ["name", "ys"],
                    "properties": {
                        "name": {"type": "string"},
                        "ys": {"type": "array", "items": {"type": "number"}},
                        "unit": {"type": "string"},
                    },
                    "additionalProperties": False,
                },
            },
            "xScale": {"enum": ["linear", "log"]},
            "yScale": {"enum": ["linear", "log", "db"]},
        },
        "additionalProperties": False,
    }
    heatmap = {
        "type": "object",
        "required": ["kind", "title", "xLabel", "yLabel", "xs", "ys", "matrix"],
        "properties": {
            "kind": {"const": "heatmap"},
            "title": {"type": "string"},
            "xLabel": {"type": "string"},
            "yLabel": {"type": "string"},
            "xs": {"type": "array", "items": {"type": "number"}},
            "ys": {"type": "array", "items": {"type": "number"}},
            "matrix": {
                "type": "array",
                "items": {"type": "array", "items": {"type": "number"}},
            },
            "unit": {"type": "string"},
            "colormap": {"type": "string"},
            "vmin": {"type": "number"},
            "vmax": {"type": "number"},
        },
        "additionalProperties": False,
    }
    bar = {
        "type": "object",
        "required": ["kind", "title", "xLabel", "yLabel", "categories", "series"],
        "properties": {
            "kind": {"const": "bar"},
            "title": {"type": "string"},
            "xLabel": {"type": "string"},
            "yLabel": {"type": "string"},
            "categories": {"type": "array", "items": {"type": "string"}},
            "series": {
                "type": "array",
                "items": {
                    "type": "object",
                    "required": ["name", "values"],
                    "properties": {
                        "name": {"type": "string"},
                        "values": {"type": "array", "items": {"type": "number"}},
                        "unit": {"type": "string"},
                    },
                    "additionalProperties": False,
                },
            },
        },
        "additionalProperties": False,
    }
    scalar = {
        "type": "object",
        "required": ["kind", "title", "rows"],
        "properties": {
            "kind": {"const": "scalar"},
            "title": {"type": "string"},
            "rows": {
                "type": "array",
                "items": {
                    "type": "object",
                    "required": ["label", "value"],
                    "properties": {
                        "label": {"type": "string"},
                        "value": {"type": ["number", "string"]},
                        "unit": {"type": "string"},
                    },
                    "additionalProperties": False,
                },
            },
        },
        "additionalProperties": False,
    }
    return {
        "$schema": "https://json-schema.org/draft/2020-12/schema",
        "title": "ChartSpec",
        "oneOf": [line, heatmap, bar, scalar],
    }


if __name__ == "__main__":
    json.dump(dump_schema(), sys.stdout, indent=2, sort_keys=True)
    sys.stdout.write("\n")

File: src/shared/chartSpec.ts
/**
 * ChartSpec IR — the wire contract between python-backend/recipe_runner.py
 * and the webview renderer in src/webview/chartSpecRenderScript.ts.
 *
 * Kept hand-mirrored from python-backend/chart_spec.py. The
 * src/test/chartSpecSchema.test.ts test spawns python -m chart_spec and
 * compares its JSON Schema output against a checked-in snapshot at
 * src/test/__snapshots__/chartSpecSchema.json — any drift between the two
 * sides surfaces there.
 */

export type ChartKind = 'line' | 'heatmap' | 'bar' | 'scalar';

export interface LineSeries {
    name: string;
    ys: number[];
    unit?: string;
}

export interface LineChart {
    kind: 'line';
    title: string;
    xLabel: string;
    yLabel: string;
    xs: number[];
    series: LineSeries[];
    xScale?: 'linear' | 'log';
    yScale?: 'linear' | 'log' | 'db';
}

export interface HeatmapChart {
    kind: 'heatmap';
    title: string;
    xLabel: string;
    yLabel: string;
    xs: number[];
    ys: number[];
    matrix: number[][];
    unit?: string;
    colormap?: string;
    vmin?: number;
    vmax?: number;
}

export interface BarSeries {
    name: string;
    values: number[];
    unit?: string;
}

export interface BarChart {
    kind: 'bar';
    title: string;
    xLabel: string;
    yLabel: string;
    categories: string[];
    series: BarSeries[];
}

export interface ScalarRow {
    label: string;
    value: number | string;
    unit?: string;
}

export interface ScalarTable {
    kind: 'scalar';
    title: string;
    rows: ScalarRow[];
}

export type ChartSpec = LineChart | HeatmapChart | BarChart | ScalarTable;

export interface RecipeRunnerResult {
    charts: ChartSpec[];
}

export function isLineChart(c: ChartSpec): c is LineChart {
    return c.kind === 'line';
}

export function isHeatmapChart(c: ChartSpec): c is HeatmapChart {
    return c.kind === 'heatmap';
}

export function isBarChart(c: ChartSpec): c is BarChart {
    return c.kind === 'bar';
}

export function isScalarTable(c: ChartSpec): c is ScalarTable {
    return c.kind === 'scalar';
}
